feat(ivy): introduce missing-injectable migration for google3

devversion opened the pull request:
Introduces a new migration schematic for adding the `@Injectable()`
decorator to provider classes which are currently not migrated. Previously
in ViewEngine, classes which are declared as providers sometimes don't
require the `@Injectable()` decorator (e.g. https://stackblitz.com/edit/angular-hpo7gw)

With Ivy, provider classes need to be explicitly decorated with the `@Injectable()`
decorator if they are declared as providers of a given module. This commit
introduces a migration schematic that adds the explicit decorator automatically
to places where the decorator is currently missing.

The migration logic is designed in a CLI devkit and TSlint agnostic way so that we
can eventually also enable this as a public CLI migration for `ng update`. This will
be handled as part of a follow-up to reiterate on console output etc.

Resolves FW-1371

devversion committed:
feat(ivy): introduce missing-injectable migration for google3

Introduces a new migration schematic for adding the "@Injectable()"
decorator to provider classes which are currently not migrated. Previously
in ViewEngine, classes which are declared as providers sometimes don't
require the "@Injectable()" decorator
(e.g. https://stackblitz.com/edit/angular-hpo7gw)

With Ivy, provider classes need to be explicitly decorated with
the "@Injectable()" decorator if they are declared as providers
of a given module. This commit introduces a migration schematic
which automatically adds the explicit decorator to places where
the decorator is currently missing.

The migration logic is designed in a CLI devkit and TSlint agnostic
way so that we can also have this migration run as part of a public
CLI migration w/ `ng update`. This will be handled as part of a follow-up to reiterate on console output etc.

Resolves FW-1371

JoostK commented on packages/core/schematics/migrations/missing-injectable/module_collector.ts:
```diff
@@ -0,0 +1,68 @@
+  private _visitNgModuleClass(node: ts.ClassDeclaration, decorator: NgDecorator) {
+    const decoratorCall = decorator.node.expression;
+
+    if (!ts.isObjectLiteralExpression(decoratorCall.arguments[0])) {
```
Could you add a test for `@NgModule()` as I think this will currently crash here.

devversion replied:
Good catch. done!

JoostK commented on packages/core/schematics/migrations/missing-injectable/module_collector.ts:
```diff
@@ -0,0 +1,68 @@
+      return;
+    }
+
+    const metadata = decoratorCall.arguments[0] as ts.ObjectLiteralExpression;
```
If you do the assignment first and then check its type using `ts.isObjectLiteralExpression` the type will be narrowed automatically:

```suggestion
const metadata = decoratorCall.arguments[0];
if (metadata === null || !ts.isObjectLiteralExpression(metadata)) {
  return;
}
```

JoostK commented on packages/core/schematics/migrations/missing-injectable/transform.ts:
```diff
@@ -0,0 +1,116 @@
+
+    const evaluatedExpr = this.partialEvaluator.evaluate(module.providersExpr);
+
+    if (!(evaluatedExpr instanceof Array)) {
```
```suggestion
    if (!Array.isArray(evaluatedExpr)) {
```

devversion replied:
What's the benefit of using `Array.isArray` here? Performance or that `isArray` covers more cases?

Switching to `isArray` for consistency anyway.. but I'd be interested :smile:

JoostK replied:
They're generally the same, only if you cross iframe boundaries will you find that the `instanceof` check may fail. Since that's not relevant in this context, consistency is the primary reason (`Array.isArray` is far more prevalent)

devversion replied:
Okay I figured it would be consistency. thanks!

JoostK commented on packages/core/schematics/migrations/missing-injectable/transform.ts:
```diff
@@ -0,0 +1,116 @@
+      } else {
+        return this._visitProviderResolvedValue(value.get('provide') !);
+      }
+    } else if (value instanceof Array) {
```
```suggestion
    } else if (Array.isArray(value)) {
```

JoostK commented on packages/core/schematics/migrations/missing-injectable/import_manager.ts:
```diff
@@ -0,0 +1,237 @@
+        continue;
+      }
+
+      if (!importStartIndex) {
```
```suggestion
      if (importStartIndex === 0) {
```

devversion replied:
Yeah. shouldn't have any effect, but explicit is always better 👍

JoostK commented on packages/core/schematics/migrations/missing-injectable/import_manager.ts:
```diff
@@ -0,0 +1,237 @@
+  addImportToSourceFile(
+      sourceFile: ts.SourceFile, symbolName: string|null, moduleName: string,
+      typeImport = false): ts.Expression {
+    const sourceDir = dirname(sourceFile.fileName);
```
I would prefer to import `path` as a namespace (but that might be just me)

devversion replied:
I'm actually a fan of importing this manually, but I don't feel too strong about it.

JoostK commented on packages/core/schematics/migrations/missing-injectable/import_manager.ts:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    if (existingImport) {
+      const propertyIdentifier = ts.createIdentifier(symbolName !);
```
Is the `symbolName !` safe here? I don't see how it's guaranteed to be set here.

devversion replied:
`existingImport` is only assigned if the `symbolName` is truthy. So this is guaranteed to be set here.

JoostK commented on packages/core/schematics/migrations/missing-injectable/import_manager.ts:
```diff
@@ -0,0 +1,237 @@
+      // specified source file and we want to add the identifiers to the existing
+      // import declaration, we need to keep track of the updated import declarations.
+      // We can't directly update the import declaration for each identifier as this
+      // would throw of the recorder offsets. We need to keep track of the new identifiers
```
```suggestion
      // would throw off the recorder offsets. We need to keep track of the new identifiers
```

JoostK commented on packages/core/schematics/migrations/missing-injectable/import_manager.ts:
```diff
@@ -0,0 +1,237 @@
+    this.getUpdateRecorder(sourceFile)
+        .addNewImport(
+            importStartIndex,
+            `\n${this.printer.printNode(ts.EmitHint.Unspecified, newImport, sourceFile)}`);
```
I would pull out the printer stuff into an intermediate var.

JoostK commented on packages/core/schematics/migrations/missing-injectable/import_manager.ts:
```diff
@@ -0,0 +1,237 @@
+
+      recorder.updateExistingImport(
+          namedBindings,
+          this.printer.printNode(ts.EmitHint.Unspecified, newNamedBindings, sourceFile));
```
Maybe pull out the printer business here as well.

JoostK commented on packages/core/schematics/migrations/missing-injectable/index.ts:
```diff
@@ -0,0 +1,120 @@
+  const rootSourceFiles = program.getRootFileNames().map(f => program.getSourceFile(f) !);
+
+  // Analyze source files by detecting all modules.
+  rootSourceFiles.forEach(sourceFile => moduleCollector.visitNode(sourceFile));
```
Are you only interested in root files? I don't see how this would fan out to all NgModules in the program.

devversion replied:
Thanks for the heads up. This is actually something I fixed for other schematics as well (see: https://github.com/angular/angular/commit/349935a4345a1f0eb66d9119d30ba3d546b6b32a) but didn't update my boilerplate for that migration.

JoostK commented on packages/core/schematics/migrations/missing-injectable/index.ts:
```diff
@@ -0,0 +1,120 @@
+    if (updateRecorders.has(sourceFile)) {
+      return updateRecorders.get(sourceFile) !;
+    }
+    const treeRecorder = tree.beginUpdate(relative(basePath, sourceFile.fileName));
```
Arent's `ts.SourceFile`s expected to have their `fileName` being an absolute path?

devversion replied:
Can you elaborate what you mean? The devkit update recorder expects paths that match the virtual filesystem tree.

JoostK replied:
Sorry, I read `resolve` instead of `relative`. My comment therefore made zero sense 😱

JoostK reviewed (commented):
Quickly went over it just now

devversion committed:
fixup! feat(ivy): introduce missing-injectable migration for google3

Address feedback

alexeagle commented on packages/core/schematics/migrations/missing-injectable/google3/noMissingInjectableRule.ts:
```diff
@@ -0,0 +1,64 @@
+/**
```
our tslint config for google3 has a block like
```
"rulesDirectory": [
    "rules",
    "../../../third_party/javascript/angular2/rc/packages/core/schematics/migrations/static-queries/google3"
  ]
```
and it's annoying to need to update this with a new directory for every angular migration.

WDYT about `packages/core/schematics/migrations/google3` containing all the `*Rule` files?

devversion replied:
Great idea. Done! Hope that helps :smile:

alexeagle commented:
Trying to dry-run this over google3, but in 14/3197 shards the v8 VM runs out of heap running this tslint check. If I'm reading the clues in the stack dump correctly, we have 4G of heap.

The input is 39MB of transitive .d.ts files and sources for a target in cloud console. I can try running the analysis on this one target locally tomorrow - you could poke around for memory leaks before that if you like.

devversion committed:
refactor(core): move google3 migration rules into single directory

Moves all google3 migration tslint rules into a single directory.
This makes it easier to wire up multiple migration rules in
google3 without having to update the rule directories each time
a new migration is available.

devversion commented:
@alexeagle Addressed your comment. Also I tried looking for memory leaks on our side. Unfortunately TSlint doesn't do a great job with managing memory when a lot of fixes are applied. We have similar issues on Material where our public migration also uses TSLint..

Reasons why TSLint is not helping here:

* No global analysis phase. We need to analyze the whole project for each `applyWithProgram` call. (I had some caching logic but the analysis phase is very fast and doesn't seem to be the culprit)
* For each replacement in a source-file. The lint rules are re-executed to find any new failures (this is a noop in our case since we don't generate new code that needs to be fixed)
* After replacements for a file have been applied, TSLint re-creates the TypeScript program. This causes significant memory pressure.
* RuleFailures have a handle to source file nodes. Meaning that these are retained once the new program is created (tried providing our own implementation of a rule failure class, but that's not the root cause)

Based on some related investigation for the Material upgrade schematics: Running migrations without TSLint for 200 source files can be done in ~10sec while with TSLint it takes ~10min (on windows)

alexeagle commented:
Update: we landed TS 3.4 in g3 yesterday, this was what caused the OOMs in the mapreduce. Syncing prior to that allowed the schematic to run to completion, I now have a CL of 218 changed g3 files, sanity checking that now.

alexeagle commented:
Found a bug: a file that had no imports at all doesn't get the new `@Injectable` import

![image](https://user-images.githubusercontent.com/47395/59710612-1eea5c00-91be-11e9-9dec-9755ba17f4fa.png)

kara commented:
A few things we should fix from the test run:

1)  We shouldn't move unrelated comments:

```ts
import a from 'b';
import x from 'y'; // some comment
```

turns into:

```ts
import {Injectable} from '@angular/core';  // some comment
import a from 'b';
import x from 'y'; 
```

but should be:

```ts
import {Injectable} from '@angular/core'; 
import a from 'b';
import x from 'y';  // some comment
```

2) We need a newline after a new import is added so comments at the top of the file stay on their own line.
```ts

/**
 * The CSS class...
```
turns into:

```ts
import {Injectable} from '@angular/core'; /**
 * The CSS class...
```

but should be: 

```ts
import {Injectable} from '@angular/core'; 
/**
 * The CSS class...
```

3) Noticing in some cases, people are using `@Inject({})` instead of `@Injectable()`. Rather than adding `@Injectable()`, seems like we should replace it.

```ts
@Inject({})
export class ParametersColorService
```

turns into

```ts
@Injectable()
@Inject({})
export class ParametersColorService
```

but should be

```ts
@Injectable()
export class ParametersColorService
```

devversion committed:
fixup! feat(ivy): introduce missing-injectable migration for google3

Insert imports after trailing comments. Add new-line after generated import at start of file

devversion committed:
fixup! feat(ivy): introduce missing-injectable migration for google3

Replace existing @Inject decorator with @Injectable

devversion commented:
@kara Feedback has been addressed. The latest two fixup commits contain the requested changes.

alexeagle commented:
Running it through tsunami again, if it looks good I'll fix up the `BUILD.bazel` files and start landing changes

alexeagle commented:
seems like the fix for when there are no imports at all tries to write two replacements at the same offset?

```
java.lang.IllegalArgumentException: import { Injectable } from "@angular/core";            
 conflicts with existing replacement @Injectable() 
```

not that important, only happens in two files so I can resolve manually


Also when replacing Inject with Injectable this left behind some unused imports, again not a big deal I already fixed them

devversion commented:
@alexeagle Thanks for testing the migration. The fact that the `Inject` import remains in the source file is expected. This is because we don't want to over engineer the migration to detect if the import is used somewhere else, or adjust logic in the update recorder to be able to remove symbols from existing imports.

Regarding the case where imports are not generated. That looks reasonable. I'll play with it. Thanks!

devversion committed:
fixup! feat(ivy): introduce missing-injectable migration for google3

Ensure imports and decorators are inserted properly if the offset is the same

alexeagle commented:
NOTE TO CARETAKER: cl/253889035 needs to be patched into the sync CL since some files were moved in this PR

kara reviewed (changes requested):
Mostly looks ready, just a few minor comments

devversion committed:
fixup! feat(ivy): introduce missing-injectable migration for google3

Address feedback from Kara

devversion commented on packages/core/schematics/test/google3/missing_injectable_rule_spec.ts:
```diff
@@ -0,0 +1,226 @@
+    const failures = linter.getResult().failures;
+
+    expect(failures.length).toBe(2);
+    expect(failures[0].getFailure()).toMatch(/Class needs to be decorated with: @Injectable\(\)/);
```
Good idea. Done!

kara reviewed (approved):
LGTM
